main/src/io/factory.hpp

```diff
@@ -31,18 +31,51 @@
 #pragma once
 
 #include <memory>
+#include <filesystem>
 
 #include "ifile_io_impl.h"
 
 namespace sphexa
 {
 
-std::unique_ptr<IFileWriter> fileWriterFactory(bool ascii, MPI_Comm comm)
+std::unique_ptr<IFileWriter> fileWriterFactory(bool ascii, MPI_Comm comm, const std::string& filePath = "",
+                                               const std::string& compression = "")
 {
     if (ascii) { return makeAsciiWriter(comm); }
-    else { return makeH5PartWriter(comm); }
+    // If the file suffix ends in ".bp", use ADIOS reader/writer.
+    // else use H5Part reader/writer, but then compression is unavailable.
+    auto suffix = std::filesystem::path(filePath).extension().string();
+    if (suffix == ".bp")
+    {
+#ifdef SPH_EXA_HAVE_ADIOS
+        return makeADIOSWriter(comm, compression);
+#endif
+        throw std::runtime_error(
+            "unsupported compression file i/o choice. BP I/O is only available with ADIOS2 enabled.\n");
+    }
+    else
+    {
+        if (compression != "")
+        {
+            throw std::runtime_error("unsupported compression file i/o choice. Output compression is only available "
+                                     "with BP file and ADIOS2 enabled.\n");
+        }
+        return makeH5PartWriter(comm);
+    }
 }
 
-std::unique_ptr<IFileReader> fileReaderFactory(bool /*ascii*/, MPI_Comm comm) { return makeH5PartReader(comm); }
+std::unique_ptr<IFileReader> fileReaderFactory(bool /*ascii*/, MPI_Comm comm, const std::string& filePath = "",
+                                               const std::string& compression = "")
+{
+    auto suffix = std::filesystem::path(filePath).extension().string();
+    if (suffix == ".bp")
+    {
+#ifdef SPH_EXA_HAVE_ADIOS
+        return makeADIOSReader(comm, compression);
```

Comment:
I think the compression settings should not be required to initalize `AdiosReader`, they should be stored in the checkpoint file if ADIOS is not already doing that. Otherwise you won't be able to read a file if you lost what settings were used.

If you have the settings available in the file, the `AdiosReader`, should be able to read any file, also uncompressed HDF5, saving you the trouble of passing two file readers to all the simulation initializers.

Reply:
That's a very good consideration and will involve some additional work. I'll leave it unresolved for the time being and revisit it after resolving the Piz Daint issue. In the meantime, to maintain simplicity in the modifications, I propose that we minimize the changes to the I/O APIs in develop branch until then.